### DISCUSSION TITLE
Could we make substreams of various resolutions unnecessary by scaling cutouts better and pull them from the main stream?

### DISCUSSION THREAD

**Discussion by brujoand** [blakeblackshear/frigate#15400] [Ask A Question] [answered]
Disclaimer: Both my current understand of how Frigate works and object detection in general might be wrong

My understanding is that frigate looks at the detect stream, and checks if pixels change. There are thresholds and masks to consider but basically you might find that pixel at coordinates 245x954 has changed, now we should figure out what it is. At this point frigate cut's out a given square around the changed point and analyses it. In the docs there is a description of how large this square will be, and how we should choose our detect resolution such that the thing we want to detect fits into this square.

For me that means having higher resolution for some of the longer distance cameras, which means I can't use the native substream but instead have to transcode it. This is currently the case for all my cameras as they only support 640*360. 

What if instead I used my tiny substream for 'change detection', but was able to configure the size of the detect square. But not have it pulled from the detect stream. Instead pull it from the big stream. 

I realize this will be confusing until some refactoring of concepts has landed but basically:
- You never need a high resolution detect stream
- The low resolution stream only needs to trigger "movement"
- We get much better resolution on the cutouts that go to image detection. If these are too high we could instead scale them down. Which would be much cheaper than transcoding the entire main stream into a lower res stream.

Am I making sense or have I completely missed something here?

**Comment by brujoand** [comment]
hm so ref https://docs.frigate.video/frigate/camera_setup#choosing-a-detect-resolution

> The ideal resolution for detection is one where the objects you want to detect fit inside the dimensions of the model used by Frigate (320x320). Frigate does not pass the entire camera frame to object detection. 

So in other words, I would have to define the size of the square "on paper". Then figure out how to scale that down to 320x320.
Which doesn't sound like a terrible idea. With this approach I could even look for squares of multiple sizes. Say I want to find both cats and trucks it could make sense to check multiple square sizes.

**Comment by NickM-27** [comment, marked as answer]
There are many problems with this approach: 

1. Streams are often not synced so you'd end up missing objects because they aren't in the high res frame
2. The only way to run detection on a frame is to decode a stream. This is already resource intensive so decoding the full res stream and the sub stream will be even worse. 

In general this is why getting cameras with good sub streams is recommended in the docs, so you can use a sub stream of 1280x720 (sweet spot in most cases) or 1920x1080 if the extra detail is needed. 

Even then you can still use the full res stream but set the detect resolution to that size in order for the GPU to be used to scale the stream down for detect.

**Comment by brujoand** [comment]
Right point 1 there would pretty much be the death blow to this approach then I guess. Thanks :+1:
